esp/services/ws_workunits/ws_workunitsHelpers.cpp

```diff
@@ -4011,6 +4011,19 @@ IFileIOStream* CWsWuFileHelper::createWUFileIOStream(IEspContext& context, const
     return createIOStreamWithFileName(zipFileNameWithPath.str(), IFOread);
 }
 
+void CWsWuFileHelper::checkFilePath(const char *file, bool UNCFileName, const char *fileType, const char *compType, const char *compName)
+{
+    StringBuffer actualPath;
+    if (UNCFileName)
```

Comment:
You are splitting the path, validating one part, and then using the orinal path.  Is there any chance the other parts can be used to hack the system.  It may be the way they are parsed they won't have at least "../" but it's worth considering.

Reply:
The other parts are machine part (url or IP) and the tail part (the part after the last PathSepChar). Any idea how to check them?

Reply:
The other parts are the machine part (url or IP) and the tail part (the part after the last PathSepChar). I haven't got an idea for the tail part. For the machine part, we may consider (in another PR?) to read the allowed machine list (with/without the allowed folder list for each machine) when the ESP is started (and when environment changes).

Reply:
If its available (c++17) you should probably run the full path through: 
 std::filesystem::canonical (https://en.cppreference.com/w/cpp/filesystem/canonical) as it will clean it up and give you the "real" path, including symlinks etc.

Reply:
I just tried:
#include <iostream>
#include <filesystem>
namespace fs = std::filesystem;

I got: error: ‘filesystem’ is not a namespace-name

Reply:
Its ultimately a wrapper around realpath:  https://pubs.opengroup.org/onlinepubs/9699919799/functions/realpath.html

Which we do use already...

Reply:
Actually looks like jfile has wrapped it in `makeAbsolutePath`?

Reply:
After using std::experimental::filesystem::path, I got:

/home/lexis/src3/HPCC-Platform/esp/services/ws_fs/ws_fsService.cpp:2800: undefined reference to `std::experimental::filesystem::v1::current_path[abi:cxx11]()'
/usr/bin/ld: /home/lexis/src3/HPCC-Platform/esp/services/ws_fs/ws_fsService.cpp:2800: undefined reference to `std::experimental::filesystem::v1::canonical(std::experimental::filesystem::v1::__cxx11::path const&, std::experimental::filesystem::v1::__cxx11::path const&)'
/usr/bin/ld: CMakeFiles/ws_fs.dir/ws_fsService.cpp.o: in function `std::experimental::filesystem::v1::__cxx11::path::path<char [16], std::experimental::filesystem::v1::__cxx11::path>(char const (&) [16])':
/usr/include/c++/9/experimental/bits/fs_path.h:212: undefined reference to `std::experimental::filesystem::v1::__cxx11::path::_M_split_cmpts()'
collect2: error: ld returned 1 exit status

Reply:
I just tried the makeAbsolutePath:
```
    StringBuffer s1, s2, s3, s4;
    makeAbsolutePath("//10.0.1.15/a/b/../c/./d/e", s1);
    makeAbsolutePath("/a/b/../c/./d/e", s2);
    makeAbsolutePath("~/a/b/../c/./d/e", s3);
    makeAbsolutePath("\\x\\y\\..\\..\\a\\b\\c", s4);
    DBGLOG("####results s1=(%s), s2=(%s), s3=(%s), s4=(%s)", s1.str(), s2.str(), s3.str(), s4.str());
```
I got: "####results s1=(//10.0.1.15/a/b/../c/./d/e), s2=(/a/b/../c/./d/e), s3=(/home/lexis/a/b/../c/./d/e), `s4=(\x\y\..\..\a\b\c)"
I am not sure it helps...

Reply:
Hmmm the comments suggest it should be handling those...   If you debug it does it hit "realpath" at anypoint?

Reply:
Yes for s2, s3, and s4.

Reply:
After adding the link_libraries(stdc++fs) to CMakeLists, the std::experimental::filesystem::canonical works. But, it does not work for a remote file.